[![Binder](http://mybinder.org/badge.svg)](http://mybinder.org:/repo/mikheyev/mucrosquamatus-selection)

# Both positive selection and genetic drift contribute to the extraordinarily rapid evolution of snake venom chemistry
- Testing the adaptive and overkill hypotheses using the Taiwan habu genome (_Protobothrops mucrosquamatus_)

## Abstract

Venoms are among the most biologically active secretions known, containing potent toxins that can serve as pharmaceutical lead structures. In particular, snake venoms consist of protein formulations that evolve rapidly, but the underlying evolutionary forces have remained controversial. Extensive evidence suggests that venom proteins are subject to strong positive selection, and that many changes are adaptive. An alternative view posits that because venoms contain numerous pharmacologically redundant components, those components could be subject to relaxed selection, and sequence evolution could be largely neutral. Yet, the two models of venom sequence evolution have ever been jointly tested. Using a combination of de novo genome sequencing, population genomics, transcriptomics, and proteomics we compare the two hypotheses in the pitviper, Protobothrops mucrosquamatus. By partitioning selective constraints and adaptive evolution in a McDonald-Kreitman-type framework, we find support for both hypotheses: venom proteins indeed experience both stronger positive selection, and lower selective constraint than other genes in the genome. Furthermore, the strength of selection may be modulated by expression level, with more abundant proteins experiencing weaker selective constraint, leading to the accumulation of more deleterious mutations. These findings show that snake venoms evolve by a combination of adaptive and neutral mechanisms, both of which explain their extraordinarily high rates of molecular evolution. In addition to positive selection, which optimizes efficacy of the venom in the short term, relaxed selective constraints for deleterious mutations can lead to more rapid turnover of individual proteins, and potentially to exploration of a larger venom phenotypic space.
